File: 647.palindromic-substrings.py
#
# @lc app=leetcode id=647 lang=python3
#
# [647] Palindromic Substrings
#

# @lc code=start
class Solution:
    def __init__(self):
        self.count = 0
    
    def countSubstrings(self, s: str) -> int:
        for i in range(len(s)):
            self.count += 1
            self.expand(s,i,i)
        for i in range(len(s)-1):
            if s[i] == s[i+1]:
                self.count += 1
                self.expand(s,i,i+1)
        return self.count    
    def expand(self,s,i,j):
        if i == 0:
            return
        if j == len(s) - 1:
            return 
        if s[i-1] == s[j+1]:
            self.count += 1
            self.expand(s,i-1,j+1)
        
# @lc code=end

File: 647.palindromic-substrings.go
/*
 * @lc app=leetcode id=647 lang=golang
 *
 * [647] Palindromic Substrings
 */
package main

import "fmt"

// @lc code=start
func countSubstrings(s string) int {
	count := 0
	for i := 0; i < len(s); i++ {
		left := i
		right := i
		// situation 1 s[i] is center of substring
		for left >= 0 && right < len(s) && s[left] == s[right] {
			count++
			left--
			right++
		}

		// situation 2 s[i] is not center of substring, but end of left part of sub string

		left = i
		right = i + 1
		for left >= 0 && right < len(s) && s[left] == s[right] {
			count++
			left--
			right++
		}
	}
	return count
}

// @lc code=end

func main() {
	fmt.Println(countSubstrings("aaa"))
}
